auth/src/main/kotlin/com/hedvig/android/auth/interceptor/AccessTokenAuthenticator.kt

```diff
@@ -0,0 +1,32 @@
+package com.hedvig.android.auth.interceptor
+
+import com.hedvig.android.auth.AuthTokenService
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import okhttp3.Authenticator
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.Route
+
+class AccessTokenAuthenticator(
+  private val authTokenService: AuthTokenService,
+) : Authenticator {
+
+  private val mutex = Mutex()
+
+  override fun authenticate(route: Route?, response: Response): Request? {
+    return runBlocking {
+      val accessToken = authTokenService.getToken()
+      mutex.withLock {
+        val newAccessToken = authTokenService.getToken()
+        if (newAccessToken != null && newAccessToken != accessToken) {
+          return@withLock response.request.withAuthorizationToken(newAccessToken.token)
+        }
+        // Still same old token, so fetch a new one.
+        val refreshedToken = authTokenService.refreshAndGetToken() ?: return@withLock null
+        response.request.withAuthorizationToken(refreshedToken.token)
+      }
+    }
+  }
```

Comment:
Took the liberty to copy this from my old project as I had shown you before. Mainly so that it defers the actual logic of refreshing the token over to the AuthTokenService, while the Authenticator itself is only responsible for locking and calling the right things, nothing more.